Makefile

```diff
@@ -0,0 +1,118 @@
+# here, else it contains an empty string. The net effect is to filter out
+# whether this current run of `make` requires a Python virtual environment.
+NEED_VENV := $(or \
+  $(findstring upgrade,$(MAKECMDGOALS)), \
+  $(findstring requirements,$(MAKECMDGOALS)), \
+  $(findstring all,$(MAKECMDGOALS)), \
+  $(findstring quick-check,$(MAKECMDGOALS)), \
+  $(findstring check,$(MAKECMDGOALS)), \
+  $(findstring test,$(MAKECMDGOALS)), \
+  $(findstring dist,$(MAKECMDGOALS)), \
+  $(findstring bdist-wheel,$(MAKECMDGOALS)), \
+  $(findstring sdist,$(MAKECMDGOALS)), \
+  $(findstring docs,$(MAKECMDGOALS)) \
+)
+ifeq ($(NEED_VENV),)
+  # None of the current goals requires a virtual environment.
+else
+  ifeq ($(origin VIRTUAL_ENV),undefined)
+    $(warning No Python virtual environment found, proceeding anyway)
+  else
+    PACKAGE_VERSION=$(shell python -c 'import package; print(package.__version__)')
+  endif
+endif
+
+# Create a virtual environment, either for Python3.10 (default) or using
+# the Python interpreter specified in the PYTHON environment variable.
+.PHONY: venv
+venv:
+	if [ ! -z "${VIRTUAL_ENV}" ]; then \
+	  echo "Found an activated Python virtual environment, exiting" && exit 1; \
+	fi
+	if [ -z "${PYTHON}" ]; then \
+	  echo "Creating virtual envirnoment in .venv/ for python3.10"; \
+	  python3.10 -m venv --upgrade-deps .venv; \
+	else \
+	  echo "Creating virtual envirnoment in .venv/ for ${PYTHON}"; \
+	  ${PYTHON} -m venv --upgrade-deps .venv; \
+	fi
+
+# Set up a newly created virtual environment. Note: pre-commit uses the
+# venv's Python interpreter, so if you've created multiple venvs then
+# pre-commit's git hooks run against the most recently set up venv.
+.PHONY: setup
+setup: upgrade
+	pre-commit install
+	pre-commit install --hook-type commit-msg
+	pre-commit install --hook-type pre-push
+
+# Install or upgrade an existing virtual environment based on the
+# package dependencies declared in setup.py.
+.PHONY: upgrade
+upgrade:
+	pip install --upgrade pip
+	pip install --editable .[hooks,dev,test,docs]
+
+# Generate a requirements.txt file containing version and integrity
+# hashes for all packages currently installed in the virtual environment.
+.PHONY: requirements
+requirements: requirements.txt
+requirements.txt:
+	echo "" > requirements.txt
+	# See also: https://github.com/peterbe/hashin/issues/139
+	for pkg in `pip list --format freeze`; do hashin --verbose $$pkg; done
+
+# Check, test, and build artifacts for this package.
+.PHONY: all
+all: check test dist docs
+
+# Run some or all checks over the package code base.
+.PHONY: quick-check check
+quick-check:
+	pre-commit run pylint --all-files
+	pre-commit run mypy --all-files
+check:
+	pre-commit run --all-files
+
+# Run all unit tests.
+.PHONY: test
+test:
+	pre-commit run pytest --hook-stage push
+
+# Build a source distribution package and a binary wheel distribution artifact.
+.PHONY: dist bdist-wheel sdist
```

Comment:
So the assumption is that `VIRTUAL_ENV` env variable is set (to `.` in the Actions workflows for example). If empty, this goal will attempt creating the release artifact at `/dist/` which does not exist.

Reply:
Correct, some goals assume that `VIRTUAL_ENV` exists and points at a sensible folder, which is why we have this guard:

https://github.com/jenstroeger/python-package-template/blob/e25b624aebfd083a38794bc970afa307093da0e0/Makefile#L21-L23

and the README mentions that:

https://github.com/jenstroeger/python-package-template/blob/e25b624aebfd083a38794bc970afa307093da0e0/UPSTREAM_README.md#L107

If we were to run the `dist` goal as part of an Action then it would fail because the `VIRTUAL_ENV` variable doesn’t exist and we haven’t created a venv in an Action either (see https://github.com/jenstroeger/python-package-template/pull/74#issuecomment-1073001587, and issue https://github.com/actions/setup-python/issues/359). Because currently we only run the `setup`, `check` and `test` goals we’re ok without that variable.

Reply:
@behnazh so… you’re saying to remove `$(VIRTUAL_ENV)` entirely and use `dist/` and `build/` in the root folder, next to this Makefile?

Reply:
> @behnazh so… you’re saying to remove `$(VIRTUAL_ENV)` entirely and use `dist/` and `build/` in the root folder, next to this Makefile?

Yes, I believe we shouldn't fiddle with the virtual environment at all, and in environments like Actions, the release should be stored in `dist/` and not `.venv/dist/`.

It might be worthwhile adding a note to the README to show where the package will be stored if venv is activated?

Reply:
Commit https://github.com/jenstroeger/python-package-template/pull/74/commits/063f6eb441d06d350b88f2cb9a0797e2c6e9c34c.